### DISCUSSION TITLE
What's your biggest hiccup with node-gtk?

### DISCUSSION THREAD

**Discussion by binyamin** [romgrk/node-gtk#266] [General]
Recently, I've begun to feel very comfortable with node-gtk. But when I first started using it, I was royally confused. I want to write a tutorial or guide on node-gtk.

So, what stops you from playing with node-gtk? What confuses you?

**Comment by baybal** [comment]
As of now, the nodejs loop, and every other JS code relying on it is laying in disuse.

So far, the no.1 impediment for practical use.

**Reply by binyamin** [reply]
I don't know what you mean by "nodejs loop". Can you provide a snippet?

**Reply by baybal** [reply]
http://docs.libuv.org/en/v1.x/design.html

**Reply by binyamin** [reply]
I was trying to ask what confuses people about node-gtk. Not what internals could be improved. Eg., "I find it confusing that Gtk doesn't use HTML", or "I can't seem to find enough docs"

**Reply by WebReflection** [reply]
most GTK doc usually speaks Python, not JS, and everything pythonic feels less natural in JS ... I don't think this is something we should fix, but it should be clear for consumer that's how it is, 'cause GJS has the same issue.

**Comment by baybal** [comment]
I would really want to hear what @bnoordhuis can say on integrating libUV with glib main loop.

**Comment by romgrk** [comment]
Fyi, I'll list here a few resources you can use to help write your tutorial:
 - [examples/gtk-4.js](https://github.com/romgrk/node-gtk/blob/master/examples/gtk-4.js): This shows the correct way to start a GTK4 application, which requires the user to spawn their own GLib event loop (where GTK3 used `gtk_main()`. With GTK4 coming out, I'll be gradually switching the default examples/recommendation to it.
 - [examples/editor.js](https://github.com/romgrk/node-gtk/blob/master/examples/editor.js): This shows a more complex GTK3 application
 - [romgrk/xedel](https://github.com/romgrk/xedel): This is an even more complex GTK3 application (soon switching to GTK4). It's interesting in that it shows how to extend & create your own components using `class MyWidget extends Gtk.Button { ... }`. The `class extends` support for GObject classes is still limited to JS world only but will eventually be integrated fully with the GLib type system. ([example](https://github.com/romgrk/xedel/blob/master/src/editor-view.js#L35))

**Comment by TheBlueOompaLoompa** [comment]
No intellisense makes it hard to write efficiently because I constantly have to look things up and make sure my conversion works, plus the errors don't usually tell me anything about what I did wrong.

**Reply by romgrk** [reply]
Good points! For the intellisense, have you tried using th ts-for-gjs project? It generates typings for gjs and node-gtk (sorry can't link I'm on mobile). Never used it yet but I'd ba happy to have feedback on it.
For the errors, can you give me examples of unhelpful messages?

**Reply by TheBlueOompaLoompa** [reply]
I'm not sure what I did that caused it to show unhelpful messages, but do you have any idea where you have to run the module generation command because I have no clue and every time I have run it the types havent showed up in typescript.

**Reply by romgrk** [reply]
After running the [install instructions](https://github.com/sammydre/ts-for-gjs#install) up to `npm install`, I see the typings being built in `./@types/node-gtk`. Not sure what you mean by unhelpful messages. I'm also not sure where to put the generated typings because I don't use typescript.

**Reply by binyamin** [reply]
@romgrk I didn't know about these typings. Theoretically, you (maintainer) could include the typedefs in the npm bundle, and that would work for autocomplete and stuff. See https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

**Comment by romgrk** [comment]
I have added just now a [template application](https://github.com/romgrk/node-gtk-template), this should facilitate bootstraping a proect.
